test/e2e/autoscale_test.go

```diff
@@ -312,22 +301,30 @@ func assertAutoscaleUpToNumPods(ctx *testContext, numPods int32) {
 	stopChan := make(chan struct{})
 	defer close(stopChan)
 
-	assertNumberOfPodsEvery(2*time.Second, ctx, errChan, stopChan, minPods, maxPods)
-
-	if err := generateTraffic(ctx, int(numPods*10), 30*time.Second); err != nil {
-		ctx.t.Fatalf("Error during scale up: %v", err)
-	}
-
-	if err := assertNumberOfPods(ctx, numPods, maxPods); err != nil {
-		errChan <- err
-	}
+	go func() {
+		if err := generateTraffic(ctx, int(numPods*10), 30*time.Second, stopChan); err != nil {
+			errChan <- err
+		}
+	}()
 
-	select {
-	case err := <-errChan:
-		ctx.t.Error(err.Error())
-	default:
-		// Success!
-	}
+	done := make(chan struct{})
+	timer := time.Tick(2 * time.Second)
+	go func() {
+		for {
+			select {
+			case <-stopChan:
+				return
+			case <-timer:
+				if err := assertNumberOfPods(ctx, minPods, maxPods); err != nil {
+					errChan <- err
+				}
+				if err := assertNumberOfPods(ctx, numPods, maxPods); err == nil {
+					close(done)
+				}
+			}
+		}
+	}()
+	<-done
```

Comment:
We now dropped the error checking and will never fail the test. You'll need to bring the `select` block back, but this time sensitive to the `done` channel.

```golang
select {
case err := <-errChan:
	ctx.t.Fatalf("Error while trying to observe scale: %v", err)
case <-done:
	// All good, we're done here.
}
```

Please check your test can still fail after a major refactor like this (for example by reversing an error check).

Reply:
should have tested it more thoroughly.
verified it fails now.
thanks for catching that.